### PR TITLE
use HTTP to enable jemalloc profile

### PR DESCRIPTION
**Pull request by YangKeao**
## What have you changed? (mandatory)

1. Add `activate_prof` and `deactivate_prof` for tikv_alloc to start and stop jemalloc profiling. These two functions `mallctl_set` the `prof.active` to true and false.

2. Add entry `/pprof/profile` in status_server. Get the `seconds` in uri query part and start profiling for seconds and stop, return the dumped profile. For example: accessing `/pprof/profile?seconds=30` will call `activate_prof` function first, and after 30 seconds a temporary file will be created to store the dumped profile. Finally the status_server return the file as response. Here are some special case:

	- Query without `seconds` will receive `BAD REQUEST`.
	- If mem-profiling and jemalloc features are not enabled (then `activate_prof` function will return an Error), ths response will be `NOT FOUND`
	- Other jemalloc error will cause `INTERNAL ERROR`
	- However, if `activate_prof` successed but `deactivate_prof` failed, no error will be returned. Maybe status_server needs a better way to handle error. (However, status_server is such a simple thing. I don't know whether )

## What are the type of the changes? (mandatory)

New feature

## How has this PR been tested? (mandatory)

I have tested it manually. Some unit tests will be added later.

After starting `tikv-server` with `--status-addr="STATUS_SERVER"`, we can access `http://STATUS_SERVER/pprof/profile?seconds=5` to dump profile for 5 seconds. We can use `jeprof` to use this result: run `jeprof --web --seconds=30 tikv-server http://STATUS_SERVER` will download profile from status server and generate a svg and open it in browser.

STATUS_SERVER above is the address of status_server and tikv-server is binary file path of `tikv-server`. Not only `--web` but also `--svg`, `--pdf` can be used for your convenience.

## Does this PR affect documentation (docs) or release note? (mandatory)

No. I didn't find docs related to status_server. Maybe a new document for status_server is needed.

## Does this PR affect tidb-ansible update? (mandatory)

No

## Refer to a related PR or issue link (optional)

#4272

### CODE REVIEW

File: src/server/status_server.rs [hunk 1]
```diff
@@ -48,19 +52,92 @@ impl StatusServer {
...
-            let mut response = Response::new(Body::empty());
-
+        let service = |req: Request<Body>| -> Box<dyn Future<Item=Response<Body>, Error=hyper::Error> + Send> {
             match (req.method(), req.uri().path()) {
                 (&Method::GET, "/metrics") => {
-                    *response.body_mut() = Body::from(dump());
+                    let response = Response::builder().body(Body::from(dump())).unwrap();
+                    Box::new(ok(response))
+                }
+                (&Method::GET, "/jeprof") => {
```
Comment (siddontang) [inline src/server/status_server.rs]: please abstract to a profile function

**Comment by siddontang** [comment]
Thanks @YangKeao 

Can you show us your manual test result? what is the output, how to use the result, etc.

**Comment by siddontang** [comment]
PTAL @brson

**Comment by siddontang** [comment]
Friendly ping @YangKeao

File: components/tikv_alloc/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+pub enum ProfError {
+    MemProfilingNotEnabled,
+    IOError(std::io::Error),
+    JemallocNotEnabled,
+    JemallocError(i32),
+}
+
+pub type ProfResult<T> = std::result::Result<T, ProfError>;
+
+impl Into<ProfError> for std::io::Error {
```
Comment (kennytm) [inline components/tikv_alloc/src/error.rs]:
```suggestion
impl From<std::io::Error> for ProfError {
```

File: components/tikv_alloc/src/jemalloc.rs [hunk 3]
```diff
@@ -89,6 +92,26 @@ mod profiling {
...
     }
 
+    pub fn activate_prof() -> ProfResult<()> {
+        unsafe {
+            if let Err(e) = jemallocator::mallctl_set(PROF_ACTIVE, true) {
+                error!("failed to activate profiling: {}", e);
+                return Err(ProfError::JemallocError(e));
+            };
+        }
+        return Ok(());
```
Comment (kennytm) [inline components/tikv_alloc/src/jemalloc.rs]:
```suggestion
        Ok(())
```

`return` is not needed for the final expression.

File: components/tikv_alloc/src/jemalloc.rs [hunk 4]
```diff
@@ -89,6 +92,26 @@ mod profiling {
...
+    }
+
+    pub fn deactivate_prof() -> ProfResult<()> {
+        unsafe {
+            if let Err(e) = jemallocator::mallctl_set(PROF_ACTIVE, false) {
+                error!("failed to deactivate profiling: {}", e);
+                return Err(ProfError::JemallocError(e));
+            };
+        }
+        return Ok(());
```
Comment (kennytm) [inline components/tikv_alloc/src/jemalloc.rs]:
```suggestion
        Ok(())
```

File: components/tikv_alloc/src/jemalloc.rs [hunk 5]
```diff
@@ -89,6 +92,26 @@ mod profiling {
         }
     }
 
+    pub fn activate_prof() -> ProfResult<()> {
+        unsafe {
+            if let Err(e) = jemallocator::mallctl_set(PROF_ACTIVE, true) {
+                error!("failed to activate profiling: {}", e);
+                return Err(ProfError::JemallocError(e));
+            };
```
Comment (kennytm) [inline components/tikv_alloc/src/jemalloc.rs]:
```suggestion
            }
```

File: components/tikv_alloc/src/jemalloc.rs [hunk 6]
```diff
@@ -89,6 +92,26 @@ mod profiling {
...
+        }
+        return Ok(());
+    }
+
+    pub fn deactivate_prof() -> ProfResult<()> {
+        unsafe {
+            if let Err(e) = jemallocator::mallctl_set(PROF_ACTIVE, false) {
+                error!("failed to deactivate profiling: {}", e);
+                return Err(ProfError::JemallocError(e));
+            };
```
Comment (kennytm) [inline components/tikv_alloc/src/jemalloc.rs]:
```suggestion
            }
```

**Comment by YangKeao** [comment]
Sorry for replying late. Run `jeprof --web --seconds=30 tikv-server http://STATUS_SERVER` can generate svg from result and open it in browser. But I am not familiar with it's output so I cannot decide whether it's correct :cry:

File: src/server/status_server.rs (hunk 1 above)
Comment (YangKeao) [reply src/server/status_server.rs]: Ok. I have abstracted them into two profile functions. One for dump profile asynchronously and the other is used for generate response from the result. `async_dump_profile` function can be placed in `tikv_alloc` crate but a timer is needed (and will add a lot of dependencies in `tikv_alloc`), so I don't do this.

**Comment by siddontang** [comment]
Great work, thanks @YangKeao, can you paste a screenshot of jeprof web output?

PTAL @BusyJay @brson

**Comment by BusyJay** [comment]
What if there are multiple requests run concurrently?

**Comment by YangKeao** [comment]
@siddontang 
jeprof

@BusyJay Jemalloc only has a global profiler, so now it cannot handle two requests at the same time ( one of them may be cut halfway because the profiler may have been closed ). I can add an Mutex to guarantee only one request is profiling. The request coming late will wait until previous one has finished. (Add a guard to start and stop profiler automatically)

File: components/tikv_alloc/src/profiler_guard.rs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+use super::{activate_prof, deactivate_prof};
+
+use futures::{Future, Poll};
+use futures_locks::{Mutex, MutexFut, MutexGuard};
+
+lazy_static! {
+    static ref PROFILER_MUTEX: Mutex<u32> = Mutex::new(0);
+}
+
+pub struct ProfilerLock(MutexFut<u32>);
```
Comment (siddontang) [inline components/tikv_alloc/src/profiler_guard.rs]: I prefer ProfGuard to keep the same with ProfResult

File: components/tikv_alloc/Cargo.toml [hunk 8]
```diff
@@ -16,6 +16,9 @@ mem-profiling = ["jemallocator/profiling"]
 [dependencies]
 libc = "0.2"
 log = "0.3"
+lazy_static = "0.2.1"
```
Comment (siddontang) [inline components/tikv_alloc/Cargo.toml]: IMO, I don't want this component depends on future crates.

Comment (YangKeao) [reply components/tikv_alloc/Cargo.toml]: `FutureMutex` is needed for running two requests without blocking whole thread. Without this I think I can only return an Error to user while another request is handling (with an atomic variable to record whether a profiler is running).

File: components/tikv_alloc/src/profiler_guard.rs (hunk 7 above)
Comment (YangKeao) [reply components/tikv_alloc/src/profiler_guard.rs]: Sorry, I don't understand what you mean.

File: components/tikv_alloc/src/profiler_guard.rs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+pub struct ProfilerGuard(MutexGuard<u32>);
+
+pub struct ProfilerLock(MutexFut<u32>);
+
+impl ProfilerLock {
+    pub fn new() -> ProfResult<ProfilerLock> {
+        match activate_prof() {
+            Ok(_) => Ok(ProfilerLock(PROFILER_MUTEX.lock())),
```
Comment (BusyJay) [inline components/tikv_alloc/src/profiler_guard.rs]: I think you should lock before enabling profile.

File: components/tikv_alloc/src/profiler_guard.rs (hunk 7 above)
Comment (BusyJay) [reply components/tikv_alloc/src/profiler_guard.rs]: He meant to rename it from `Profiler*` to `Prof*`.

File: components/tikv_alloc/Cargo.toml (hunk 8 above)
Comment (siddontang) [reply components/tikv_alloc/Cargo.toml]: As jay suggestion, I also prefer removing this to TiKV itself, not in this crate.

It is strange to depend on the futures crate.

**Comment by YangKeao** [comment]
@BusyJay Thanks for your suggestion :smile: . `profiler_guard` was moved to `server/status_server` as a mod now.

File: components/tikv_alloc/src/default.rs [hunk 10]
```diff
@@ -9,3 +10,11 @@ pub fn dump_prof(_path: Option<&str>) {}
 pub fn fetch_stats() -> io::Result<Option<AllocStats>> {
     Ok(None)
 }
+
+pub fn activate_prof() -> ProfResult<()> {
+    Err(ProfError::JemallocNotEnabled)
```
Comment (siddontang) [inline components/tikv_alloc/src/default.rs]: A little strange that the common function just returns a Jemalloc error.

Comment (BusyJay) [reply components/tikv_alloc/src/default.rs]: Better use `ProfileDisabled`. Actually tcmalloc also support profiling.

File: src/server/status_server.rs [hunk 11]
```diff
@@ -41,26 +87,104 @@ impl StatusServer {
...
+                    .body(Body::empty())
+                    .unwrap();
+                return Box::new(ok(response));
+            }
+        };
+
+        Box::new(
+            Self::dump_prof(seconds)
+                .and_then(|buf| {
+                    let response = Response::builder().body(buf.into()).unwrap();
```
Comment (brson) [inline src/server/status_server.rs]: Please set the "content-length" and "content-type" headers. The header constants are in `http::header` and the mime type is in the `mime` crate. I don't _think_ the response builder will automatically set the "content-length" from the body.

**Comment by brson** [comment]
Does this patch still require someone to set `MALLOC_CONF=prof:true` to work? I was under the impression that enabling profiling was a _3 step_ process: build with --enable-profile, set `opt.prof`, activate `opt.prof.active`.

File: components/tikv_alloc/src/error.rs [hunk 12]
```diff
@@ -0,0 +1,15 @@
+#[derive(Debug)]
+pub enum ProfError {
+    MemProfilingNotEnabled,
+    IOError(std::io::Error),
+    JemallocNotEnabled,
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply components/tikv_alloc/src/error.rs]: Yes, it was used in default.rs. But according to BusyJay's suggestion, tcmalloc profiling will also be supported in the future, `JemallocNotEnabled` is not a suitable error for that situation.

So I will remove this variant.

File: src/server/status_server.rs [hunk 13]
```diff
@@ -1,17 +1,63 @@
...
+                Ok(_) => Ok(ProfLock(guard)),
+                Err(e) => Err(e),
+            }
+        }
+    }
+
+    impl Drop for ProfGuard {
+        fn drop(&mut self) {
+            match deactivate_prof() {
+                _ => {} // TODO: handle error here
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply src/server/status_server.rs]: This error has been logged in `deactivate_prof` function. Should I log it twice?

[the remark this answers is not in the document]
Comment (brson) [reply src/server/status_server.rs]: Ah, no that's fine.

File: src/server/status_server.rs [hunk 14]
```diff
@@ -41,26 +87,104 @@ impl StatusServer {
...
+                        })
+                        .map_err(|e| -> ProfError { e.into() })
+                })
+        }))
+    }
+
+    pub fn dump_prof_to_resp(
+        req: Request<Body>,
+    ) -> Box<dyn Future<Item = Response<Body>, Error = hyper::Error> + Send> {
+        let url = url::Url::parse(&format!("http://host{}", req.uri().to_string())).unwrap(); // Add scheme and host to parse query
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply src/server/status_server.rs]: Thanks for your suggestion! I have been looking for a better way to do this for a long time. :laughing:

**Comment by siddontang** [comment]
@brson 

> Does this patch still require someone to set MALLOC_CONF=prof:true to work?

I think we don't need this if we build with profiling feature now, because the HTTP can enable/disable automatically.

**Comment by YangKeao** [comment]
@siddontang No. Setting `prof` to be true is also required now. Because HTTP request only turns on/off `prof_active` without caring about `prof`.

Should HTTP request also auto turn on and off `prof`?

**Comment by YangKeao** [comment]
@brson Second was used as parameters because `jeprof` uses seconds.

For example `jeprof --web --seconds=30 target/debug/tikv-server http://127.0.0.1:8079` will send a request to http://127.0.0.1:8079/pprof/profile?seconds=30

**Comment by siddontang** [comment]
@YangKeao 

> Setting prof to be true is also required now. Because HTTP request only turns on/off prof_active without caring about prof.

seem it is not convenient to pass this env at runtime.  can we set `MALLOC_CONF=prof:true` directly when we start TiKV with profiling feature?

But we should ensure setting `prof:true` can't affect performance at first.

**Comment by YangKeao** [comment]
Is there anything more I can do for this PR ?

**Comment by brson** [comment]
> seem it is not convenient to pass this env at runtime. can we set `MALLOC_CONF=prof:true` directly when we start TiKV with profiling feature?

@siddontang 

Yes it is technically possible. See the "tuning" section [here](http://jemalloc.net/jemalloc.3.html#opt.prof). I don't know how complex it is to feed that configuration down through the various intermediary crates all the way to the jemalloc build system. I don't know if it will add additional overhead beyond what --enable-profiling does, but I suspect if it does it's mostly memory overhead, not instruction overhead, since it's a run-time option.

It looks to me like the most viable way to set "opt.prof" automatically is to create a statically initialized extern c-string, `malloc_conf`, containing its value. That could be done within tikv_alloc via cfgs. Though jemalloc-sys may provide some way to set --with-malloc-conf conditionally, I can't think offhand how it would do it, and haven't looked it up.

I don't think we necessarily have to solve that problem in this PR though.

> @siddontang Is there anything more I can do for this PR? I have no idea whether `prof:true` will affect its performance ( I think benchmark is needed to decide but I don't have enough computing resource to run benchmark for it )

@YangKeao sorry for the long delay's here. I've given it a quick re-review and I think it is ok. Even if not, it's way off any critical path so any lingering bugs aren't a major problem.

I'm ok with merging, as long as we file a bug for the follow up to figure out how to set `opt.prof` automatically.

Let's see what @siddontang says though.

**Comment by brson** [comment]
I went ahead and fixed the merge conflict.

File: src/server/status_server.rs [hunk 15]
```diff
@@ -41,26 +87,131 @@ impl StatusServer {
...
+                Err(_) => {
+                    let response = Response::builder()
+                        .status(StatusCode::BAD_REQUEST)
+                        .body(Body::empty())
+                        .unwrap();
+                    return Box::new(ok(response));
+                }
+            },
+            None => {
+                let response = Response::builder()
```
Comment (BusyJay) [inline src/server/status_server.rs]: Can it be set to a default value like 60 seconds instead? What's the behavior of TiDB?

Comment (YangKeao) [reply src/server/status_server.rs]:
```
sec, err := strconv.ParseInt(r.FormValue("seconds"), 10, 64)
if sec <= 0 || err != nil {
    sec = 10
} 
```
The default value in TiDB is 10 seconds.

Comment (YangKeao) [reply src/server/status_server.rs]: Now,  if `seconds` query parameter is not passed, the value will be set to 10 seconds. If it's not a integer, an error will be returned.

**Comment by siddontang** [comment]
hello @YangKeao 

do you mind adding a comment that we must export `MALLOC_CONF=prof:true` when using this feature in the following PR? 

I will merge this, thank you very much.

**Pull request merged**